### PR TITLE
Refactoring to `OpaqueTyOrigin`

### PR DESCRIPTION
**Pull request by compiler-errors**
Pulled out of a larger PR that uses these changes to do cross-crate encoding of opaque origin, so we can use them for edition 2024 migrations. These changes should be self-explanatory on their own, tho 😄

### CODE REVIEW

File: src/tools/clippy/clippy_lints/src/manual_async_fn.rs [hunk 1]
```diff
@@ -105,7 +105,7 @@ fn future_trait_ref<'tcx>(
     cx: &LateContext<'tcx>,
     ty: &'tcx Ty<'tcx>,
 ) -> Option<(&'tcx TraitRef<'tcx>, Vec<LifetimeName>)> {
-    if let TyKind::OpaqueDef(item_id, bounds, false) = ty.kind
```
Comment (compiler-errors) [inline src/tools/clippy/clippy_lints/src/manual_async_fn.rs:108]: This is the only change to semantics in clippy, but I don't think we need to explicitly opt out AFIT here.

File: compiler/rustc_privacy/src/lib.rs [hunk 2]
```diff
@@ -636,8 +636,7 @@ impl<'tcx> EmbargoVisitor<'tcx> {
 impl<'tcx> Visitor<'tcx> for EmbargoVisitor<'tcx> {
     fn visit_item(&mut self, item: &'tcx hir::Item<'tcx>) {
         if self.impl_trait_pass
-            && let hir::ItemKind::OpaqueTy(opaque) = item.kind
-            && !opaque.in_trait
```
Comment (compiler-errors) [inline compiler/rustc_privacy/src/lib.rs:640]: I'm not sure why we were opting out `in_trait` from the embargo visitor, but I didn't see any change to behavior here.

**Comment by compiler-errors** [comment]
Hold up not ready yet

@rustbot author

**Comment by compiler-errors** [comment]
ok, I think it's ready

@rustbot ready

File: src/tools/clippy/clippy_lints/src/manual_async_fn.rs (hunk 1 above)
Comment (flip1995) [reply src/tools/clippy/clippy_lints/src/manual_async_fn.rs:108]: I don't think this is doing anything right now. We don't have tests for traits for this lint. I added one and changing `false`->`_` in the current implementation doesn't change the behavior there. So removing this should be fine :+1:

Comment (compiler-errors) [reply src/tools/clippy/clippy_lints/src/manual_async_fn.rs:108]: Thank you for checking! ❤️

**Comment by compiler-errors** [comment]
bors wtf lol

**Pull request merged**